src/App.js

```diff
@@ -21,45 +21,93 @@ const generateSquares = () => {
       currentId += 1;
     }
   }
-
   return squares;
 };
 
 const App = () => {
-  // This starts state off as a 2D array of JS objects with
-  // empty value and unique ids.
+  // This starts state off as a 2D array of JS objects with empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+  const [winner, setWinner] = useState(null);
 
-  // Wave 2
-  // You will need to create a method to change the square
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+  const setSquareValue = (id) => {
+    const newSquares = [...squares];
 
-  const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    for (let row of newSquares) {
+      for (let square of row) {
+        if (square.id === id) {
+          if (square.value === '') {
+            square.value = currentPlayer;
+            if (currentPlayer === PLAYER_1) {
+              setCurrentPlayer(PLAYER_2);
+            } else {
+              setCurrentPlayer(PLAYER_1);
+            }
+          }
+        }
+      }
+      checkForWinner(newSquares);
+      setSquares(newSquares);
+    }
   };
 
-  const resetGame = () => {
-    // Complete in Wave 4
+  const checkForWinner = (squares) => {
+    let newWinner;
+    let i = 0;
+    while (i < 3) {
```

Comment:
A while loop works, but can be easy to introduce a bug. Consider using a for loop if you're looping a set number of times.